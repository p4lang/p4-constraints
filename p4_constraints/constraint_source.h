// Copyright 2020 The P4-Constraints Authors
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#ifndef THIRD_PARTY_P4LANG_P4_CONSTRAINTS_P4_CONSTRAINTS_CONSTRAINT_SOURCE_H_
#define THIRD_PARTY_P4LANG_P4_CONSTRAINTS_P4_CONSTRAINTS_CONSTRAINT_SOURCE_H_

#include <string>

#include "p4_constraints/ast.pb.h"

namespace p4_constraints {

// Convenient struct of source information for quoting.
struct ConstraintSource {
  std::string constraint_string;
  ast::SourceLocation constraint_location;
};

}  // namespace p4_constraints

#endif  // THIRD_PARTY_P4LANG_P4_CONSTRAINTS_P4_CONSTRAINTS_CONSTRAINT_SOURCE_H_
