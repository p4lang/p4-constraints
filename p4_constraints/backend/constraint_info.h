/*
 * Copyright 2020 The P4-Constraints Authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

// Data structure containing all info required for constraint checking.
//
// ConstraintInfo is a data structure containing all information required for
// checking that a table entry satisfies the constraints specified in the P4
// program. ConstraintInfo can be parsed from a P4Info protobuf.

#ifndef P4_CONSTRAINTS_BACKEND_CONSTRAINT_INFO_H_
#define P4_CONSTRAINTS_BACKEND_CONSTRAINT_INFO_H_

#include <string>
#include <utility>
#include <vector>

#include "absl/container/flat_hash_map.h"
#include "absl/status/statusor.h"
#include "absl/types/variant.h"
#include "p4/config/v1/p4info.pb.h"
#include "p4_constraints/ast.pb.h"
#include "p4_constraints/constraint_source.h"

namespace p4_constraints {

struct KeyInfo {
  uint32_t id;       // Same as MatchField.id in p4info.proto.
  std::string name;  // Same as MatchField.name in p4info.proto.

  // Key type specifying how many bits to match on and how, e.g. Ternary<16>.
  // Derives from MatchField.match_type and MatchField.bitwidth in p4info.proto.
  ast::Type type;
};

struct TableInfo {
  uint32_t id;       // Same as Table.preamble.id in p4info.proto.
  std::string name;  // Same as Table.preamble.name in p4info.proto.

  // An optional constraint (aka entry_restriction) on table entries.
  absl::optional<ast::Expression> constraint;
  // If member `constraint` is present, this captures its source. Aribitray
  // otherwise
  ConstraintSource constraint_source;

  // Maps from key IDs/names to KeyInfo.
  // Derives from Table.match_fields in p4info.proto.
  absl::flat_hash_map<uint32_t, KeyInfo> keys_by_id;
  absl::flat_hash_map<std::string, KeyInfo> keys_by_name;
};

// Contains all information required for constraint checking.
// Technically, a map from table IDs to TableInfo.
using ConstraintInfo = absl::flat_hash_map<uint32_t, TableInfo>;

// Translates P4Info to ConstraintInfo.
//
// Parses all tables and their constraint annotations into an in-memory
// representation suitable for constraint checking. Returns parsed
// representation, or an error statuses if parsing fails.
absl::StatusOr<ConstraintInfo> P4ToConstraintInfo(
    const p4::config::v1::P4Info& p4info);

// Table entry metadata accessible in the constraint language, e.g. priority.
struct MetadataInfo {
  std::string name;
  ast::Type type;
};

// Returns information for a given metadata name, std::nullopt for unknown
// metadata.
std::optional<MetadataInfo> GetMetadataInfo(absl::string_view metadata_name);

}  // namespace p4_constraints

#endif  // P4_CONSTRAINTS_BACKEND_CONSTRAINT_INFO_H_
