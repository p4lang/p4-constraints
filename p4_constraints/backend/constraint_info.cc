// Copyright 2020 The P4-Constraints Authors
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include "p4_constraints/backend/constraint_info.h"

#include <stdint.h>

#include <string>
#include <utility>
#include <vector>

#include "absl/container/flat_hash_map.h"
#include "absl/status/status.h"
#include "absl/status/statusor.h"
#include "absl/strings/str_cat.h"
#include "absl/strings/str_join.h"
#include "absl/strings/string_view.h"
#include "absl/strings/strip.h"
#include "absl/types/optional.h"
#include "gutils/status_macros.h"
#include "p4/config/v1/p4info.pb.h"
#include "p4_constraints/ast.pb.h"
#include "p4_constraints/backend/type_checker.h"
#include "p4_constraints/constraint_source.h"
#include "p4_constraints/frontend/parser.h"
#include "re2/re2.h"

namespace p4_constraints {

namespace {

using p4::config::v1::MatchField;
using p4::config::v1::Table;

absl::StatusOr<absl::optional<ConstraintSource>> ExtractConstraint(
    const Table& table) {
  // We expect .p4 files to have the following format:
  // ```p4
  //   @file(__FILE__)              // optional
  //   @line(__LINE__)              // optional
  //   @entry_restriction("
  //      <the actual constraint>
  //   ")
  //   table foo { ... }
  // ```
  // The @file/@line annotations are optional and intended for debugging/testing
  // only; they allows us to give error messages that quote the source code.
  const RE2 file_annotation = {R"RE(@file[(]"([^"]*)"[)])RE"};
  const RE2 line_annotation = {R"RE(@line[(](\d+)[)])RE"};
  const RE2 constraint_annotation = {R"RE(@entry_restriction)RE"};

  absl::string_view constraint_string = "";
  ast::SourceLocation constraint_location;
  int line = 0;
  for (absl::string_view annotation : table.preamble().annotations()) {
    if (RE2::Consume(&annotation, file_annotation,
                     constraint_location.mutable_file_path()))
      continue;
    if (RE2::Consume(&annotation, line_annotation, &line)) continue;
    if (RE2::Consume(&annotation, constraint_annotation)) {
      constraint_string = annotation.data();
      break;
    }
  }

  if (constraint_string.empty()) return absl::nullopt;

  constraint_location.set_line(line);
  if (constraint_location.file_path().empty()) {
    constraint_location.set_table_name(table.preamble().name());
  }

  if (!absl::ConsumePrefix(&constraint_string, "(\"") ||
      !absl::ConsumeSuffix(&constraint_string, "\")")) {
    return gutils::InvalidArgumentErrorBuilder(GUTILS_LOC)
           << "In table " << table.preamble().name() << ":\n"
           << "Syntax error: @entry_restriction must be enclosed in "
              "'(\"' and '\")'";
  }
  return ConstraintSource{
      .constraint_string = std::string(constraint_string),
      .constraint_location = constraint_location,
  };
}

absl::StatusOr<ast::Type> ParseKeyType(const MatchField& key) {
  ast::Type type;
  switch (key.match_case()) {
    case MatchField::kMatchType:
      switch (key.match_type()) {
        case MatchField::EXACT:
          type.mutable_exact()->set_bitwidth(key.bitwidth());
          return type;
        case MatchField::TERNARY:
          type.mutable_ternary()->set_bitwidth(key.bitwidth());
          return type;
        case MatchField::LPM:
          type.mutable_lpm()->set_bitwidth(key.bitwidth());
          return type;
        case MatchField::RANGE:
          type.mutable_range()->set_bitwidth(key.bitwidth());
          return type;
        case MatchField::OPTIONAL:
          type.mutable_optional_match()->set_bitwidth(key.bitwidth());
          return type;
        default:
          return gutils::InvalidArgumentErrorBuilder(GUTILS_LOC)
                 << "match key of invalid MatchType: "
                 << MatchField::MatchType_Name(key.match_type());
      }
    case MatchField::kOtherMatchType:
      *type.mutable_unsupported()->mutable_name() = key.other_match_type();
      return type;
    default:
      return gutils::InternalErrorBuilder(GUTILS_LOC)
             << "unknown MatchField.match case: " << key.match_case();
  }
}

absl::StatusOr<TableInfo> ParseTableInfo(const Table& table) {
  absl::flat_hash_map<uint32_t, KeyInfo> keys_by_id;
  absl::flat_hash_map<std::string, KeyInfo> keys_by_name;

  for (const MatchField& key : table.match_fields()) {
    ASSIGN_OR_RETURN(const ast::Type type, ParseKeyType(key));
    const KeyInfo key_info{.id = key.id(), .name = key.name(), .type = type};
    if (!keys_by_id.insert({key_info.id, key_info}).second) {
      return gutils::InvalidArgumentErrorBuilder(GUTILS_LOC)
             << "table " << table.preamble().name()
             << " has duplicate key: " << key.DebugString();
    }
    if (!keys_by_name.insert({key_info.name, key_info}).second) {
      return gutils::InvalidArgumentErrorBuilder(GUTILS_LOC)
             << "table " << table.preamble().name()
             << " has duplicate key: " << key.DebugString();
    }
  }

  ASSIGN_OR_RETURN(absl::optional<ConstraintSource> constraint_source,
                   ExtractConstraint(table));

  absl::optional<ast::Expression> constraint = absl::nullopt;
  if (constraint_source.has_value()) {
    ASSIGN_OR_RETURN(constraint, ParseConstraint(*constraint_source));
  }

  TableInfo table_info{
      .id = table.preamble().id(),
      .name = table.preamble().name(),
      .constraint = constraint,
      .constraint_source = constraint_source.value_or(ConstraintSource()),
      .keys_by_id = keys_by_id,
      .keys_by_name = keys_by_name,
  };

  // Type check constraint.
  if (table_info.constraint.has_value()) {
    RETURN_IF_ERROR(InferAndCheckTypes(&*table_info.constraint, table_info));
  }

  return table_info;
}

}  // namespace

std::optional<AttributeInfo> GetAttributeInfo(
    absl::string_view attribute_name) {
  // ArbitraryInt ast type.
  ast::Type arbitrary_int;
  arbitrary_int.mutable_arbitrary_int();

  if (attribute_name == "priority") {
    return AttributeInfo{.name = "priority", .type = arbitrary_int};
  }

  // Unknown attribute.
  return absl::nullopt;
}

absl::StatusOr<ConstraintInfo> P4ToConstraintInfo(
    const p4::config::v1::P4Info& p4info) {
  // Allocate output.
  absl::flat_hash_map<uint32_t, TableInfo> info;
  std::vector<absl::Status> errors;

  for (const Table& table : p4info.tables()) {
    absl::StatusOr<TableInfo> table_info = ParseTableInfo(table);
    if (!table_info.ok()) {
      errors.push_back(table_info.status());
    } else if (!info.insert({table.preamble().id(), *table_info}).second) {
      errors.push_back(gutils::InvalidArgumentErrorBuilder(GUTILS_LOC)
                       << "duplicate table: " << table.DebugString());
    }
  }
  if (errors.empty()) return info;
  return gutils::InvalidArgumentErrorBuilder(GUTILS_LOC)
         << "P4Info to constraint info translation failed with the following "
            "errors:\n- "
         << absl::StrJoin(errors, "\n- ",
                          [](std::string* out, const absl::Status& status) {
                            absl::StrAppend(out, status.message(), "\n");
                          });
}

}  // namespace p4_constraints
