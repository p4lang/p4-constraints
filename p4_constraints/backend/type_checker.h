/*
 * Copyright 2020 The P4-Constraints Authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

// The type checker infers and checks the types of expressions.

#ifndef P4_CONSTRAINTS_BACKEND_TYPE_CHECKER_H_
#define P4_CONSTRAINTS_BACKEND_TYPE_CHECKER_H_

#include "p4_constraints/ast.pb.h"
#include "p4_constraints/backend/constraint_info.h"
#include "util/status.h"

namespace p4_constraints {

// Type checks the given expression, returning an OkStatus if type checking
// succeeds or an InvalidInput Status otherwise.
//
// Note that this is a side-effecting operation that may mutates the given
// expression in two ways:
// - It may mutate the type of the expression and its subexpressions, filling in
//   the correct types.
// - It may insert type-casts, making implicit casts explicit.
//
// Upon successful completion of this function, the given expression is
// guaranteed to contain no ast::Type::Unknown/Unsupported types.
//
// This function is idempotent, meaning that if it is called twice in a row,
// the function returns the same result each time and does not further mutate
// its input during the second call.
util::Status InferAndCheckTypes(ast::Expression* expr,
                                const TableInfo& table_info);

}  // namespace p4_constraints

#endif  // P4_CONSTRAINTS_BACKEND_TYPE_CHECKER_H_
