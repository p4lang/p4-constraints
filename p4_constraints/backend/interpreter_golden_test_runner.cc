// Copyright 2020 The P4-Constraints Authors
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

// Generates file for golden testing of `ReasonEntryViolatesConstraint` in
// interpreter.cc. Expected output is `interpreter_golden_test_runner.expected`

#include "absl/log/check.h"
#include "absl/status/statusor.h"
#include "absl/strings/str_join.h"
#include "gutils/ordered_map.h"
#include "gutils/parse_text_proto.h"
#include "gutils/status_macros.h"
#include "p4_constraints/backend/constraint_info.h"
#include "p4_constraints/backend/interpreter.h"
#include "p4_constraints/backend/type_checker.h"
#include "p4_constraints/constraint_source.h"
#include "p4_constraints/frontend/parser.h"

namespace p4_constraints {
namespace internal_interpreter {

using ::gutils::ParseTextProtoOrDie;
using ::p4_constraints::ast::Expression;
using ::p4_constraints::ast::SourceLocation;
using ::p4_constraints::ast::Type;

// A test case for `ReasonEntryViolatesConstraint` function.
struct TestCase {
  std::string constraint;
  p4::v1::TableEntry table_entry;
  bool source_file_available;
};

absl::StatusOr<ConstraintInfo> MakeConstraintInfo(TestCase test_case) {
  const Type kExact32 = ParseTextProtoOrDie<Type>("exact { bitwidth: 32 }");
  const Type kTernary32 = ParseTextProtoOrDie<Type>("ternary { bitwidth: 32 }");
  const Type kLpm32 = ParseTextProtoOrDie<Type>("lpm { bitwidth: 32 }");
  const Type kRange32 = ParseTextProtoOrDie<Type>("range { bitwidth: 32 }");
  const Type kOptional32 =
      ParseTextProtoOrDie<Type>("optional_match { bitwidth: 32 }");
  const TableInfo kTableInfo{
      .id = 1,
      .name = "golden_table",
      .keys_by_id =
          {
              {1, {1, "exact32", kExact32}}, {2, {2, "ternary32", kTernary32}},
              // For testing purposes, fine to omit the other keys here.
          },
      .keys_by_name = {
          {"exact32", {1, "exact32", kExact32}},
          {"ternary32", {2, "ternary32", kTernary32}},
          {"lpm32", {3, "lpm32", kLpm32}},
          {"range32", {4, "range32", kRange32}},
          {"optional32", {5, "optional32", kOptional32}},
      }};

  ConstraintSource source{
      .constraint_string = test_case.constraint,
      .constraint_location = SourceLocation(),
  };
  if (test_case.source_file_available) {
    *source.constraint_location.mutable_file_path() = "golden_test.p4";
  } else {
    *source.constraint_location.mutable_table_name() = kTableInfo.name;
  }

  TableInfo table_info = kTableInfo;
  table_info.constraint_source = source;
  ASSIGN_OR_RETURN(Expression expression, ParseConstraint(source));
  CHECK_OK(InferAndCheckTypes(&(expression), kTableInfo));
  table_info.constraint = expression;
  return ConstraintInfo{{table_info.id, table_info}};
}

// NOTE: Test cases only define the "exact" field for brevity but kTableInfo is
// composed of several keys. These undeclared keys are implicitly set to hold
// "catchall" values (i.e. range=[min,max], ternary=*).
std::vector<TestCase> TestCases() {
  std::vector<TestCase> test_cases;
  test_cases.push_back(TestCase{
      .constraint = "true;",
      .table_entry = ParseTextProtoOrDie<p4::v1::TableEntry>(R"pb(
        table_id: 1
        match {
          field_id: 1
          exact { value: "\012" }
        }
      )pb"),
      .source_file_available = false,
  });

  test_cases.push_back(TestCase{
      .constraint = "exact32::value != 10 || exact32::value == 10;",
      .table_entry = ParseTextProtoOrDie<p4::v1::TableEntry>(R"pb(
        table_id: 1
        match {
          field_id: 1
          exact { value: "\012" }
        }
      )pb"),
      .source_file_available = false,
  });

  test_cases.push_back(TestCase{
      .constraint = "exact32::value > 6 && exact32::value < 5;",
      .table_entry = ParseTextProtoOrDie<p4::v1::TableEntry>(R"pb(
        table_id: 1
        match {
          field_id: 1
          exact { value: "\012" }
        }
      )pb"),
      .source_file_available = false,
  });

  test_cases.push_back(TestCase{
      .constraint = "exact32::value > 5;\n"
                    "exact32::value < 20;\n"
                    "exact32::value == 14;",
      .table_entry = ParseTextProtoOrDie<p4::v1::TableEntry>(R"pb(
        table_id: 1
        match {
          field_id: 1
          exact { value: "\012" }
        }
      )pb"),
      .source_file_available = false,
  });

  test_cases.push_back(TestCase{
      .constraint = "exact32::value > 0;\n"
                    "exact32::value > 7 || exact32::value == 5;\n"
                    "exact32::value == 9;",
      .table_entry = ParseTextProtoOrDie<p4::v1::TableEntry>(R"pb(
        table_id: 1
        match {
          field_id: 1
          exact { value: "\012" }
        }
      )pb"),
      .source_file_available = true,
  });

  test_cases.push_back(TestCase{
      .constraint = "exact32::value > 0;\n"
                    "exact32::value < 42;\n"
                    "exact32::value < 20 -> exact32::value == 14;",
      .table_entry = ParseTextProtoOrDie<p4::v1::TableEntry>(R"pb(
        table_id: 1
        match {
          field_id: 1
          exact { value: "\012" }
        }
      )pb"),
      .source_file_available = true,
  });

  test_cases.push_back(TestCase{
      .constraint = "exact32::value == 1 || exact32::value == 2;\n"
                    "!(exact32::value == 10 -> exact32::value == 10);\n"
                    "exact32::value == 3 || exact32::value == 4;",
      .table_entry = ParseTextProtoOrDie<p4::v1::TableEntry>(R"pb(
        table_id: 1
        match {
          field_id: 1
          exact { value: "\012" }
        }
      )pb"),
      .source_file_available = true,
  });

  test_cases.push_back(TestCase{
      .constraint = "exact32::value == 80 || ternary32::value == 3096;\n"
                    "ternary32::mask == 255 && exact32::value == 3;",
      .table_entry = ParseTextProtoOrDie<p4::v1::TableEntry>(R"pb(
        table_id: 1
        match {
          field_id: 1
          exact { value: "\012" }
        }
        match {
          field_id: 2
          ternary { value: "\052" mask: "\100" }
        }
      )pb"),
      .source_file_available = false,
  });

  test_cases.push_back(TestCase{
      .constraint =
          "(false || false) && (!(true -> true) && (false || false));",
      .table_entry = ParseTextProtoOrDie<p4::v1::TableEntry>(R"pb(
        table_id: 1
        match {
          field_id: 1
          exact { value: "\012" }
        }
      )pb"),
      .source_file_available = true,
  });

  return test_cases;
}

std::string EntryToString(const TableEntry& entry) {
  std::string key_info = absl::StrJoin(
      gutils::Ordered(entry.keys), "\n", [](std::string* out, auto pair) {
        absl::StrAppend(out, "Key:\"", pair.first,
                        "\" -> Value: ", EvalResultToString(pair.second));
      });
  return absl::StrFormat(
      "Table Name: \"%s\"\n"
      "Priority:%d\n"
      "%s\n",
      entry.table_name, entry.priority, key_info);
}

absl::Status main() {
  for (const TestCase& test_case : TestCases()) {
    ASSIGN_OR_RETURN(ConstraintInfo constraint_info,
                     MakeConstraintInfo(test_case));
    ASSIGN_OR_RETURN(TableEntry input,
                     ParseEntry(test_case.table_entry, constraint_info.at(1)));
    std::cout << "### ReasonEntryViolatestConstraint Test ###################\n"
              << "=== INPUT ===\n"
              << "--- Constraint ---\n"
              << test_case.constraint << "\n"
              << "--- Table Entry ---\n"
              << EntryToString(input) << "\n";

    absl::StatusOr<std::string> result =
        ReasonEntryViolatesConstraint(test_case.table_entry, constraint_info);
    if (result.ok()) {
      std::cout << "=== OUTPUT ===\n"
                << (result->empty() ? "<empty>\n" : *result) << "\n";
    } else {
      std::cout << "=== ERROR ===\n" << result.status();
    }
  }
  return absl::OkStatus();
}

}  // namespace internal_interpreter
}  // namespace p4_constraints

int main() { CHECK_OK(p4_constraints::internal_interpreter::main()); }
