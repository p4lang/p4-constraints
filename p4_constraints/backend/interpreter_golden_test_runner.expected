### ReasonEntryViolatestConstraint Test ###################
=== INPUT ===
--- Constraint ---
true;
--- Table Entry ---
Table Name: "golden_table"
Priority:0
Key:"exact32" -> Value: Exact{.value = 10}
Key:"lpm32" -> Value: Lpm{.value = 0, .prefix_length = 0}
Key:"optional32" -> Value: Ternary{.value = 0, .mask = 0}
Key:"range32" -> Value: Range{.low = 0, .high = 4294967295}
Key:"ternary32" -> Value: Ternary{.value = 0, .mask = 0}

=== OUTPUT ===
<empty>

### ReasonEntryViolatestConstraint Test ###################
=== INPUT ===
--- Constraint ---
exact32::value != 10 || exact32::value == 10;
--- Table Entry ---
Table Name: "golden_table"
Priority:0
Key:"exact32" -> Value: Exact{.value = 10}
Key:"lpm32" -> Value: Lpm{.value = 0, .prefix_length = 0}
Key:"optional32" -> Value: Ternary{.value = 0, .mask = 0}
Key:"range32" -> Value: Range{.low = 0, .high = 4294967295}
Key:"ternary32" -> Value: Ternary{.value = 0, .mask = 0}

=== OUTPUT ===
<empty>

### ReasonEntryViolatestConstraint Test ###################
=== INPUT ===
--- Constraint ---
exact32::value > 6 && exact32::value < 5;
--- Table Entry ---
Table Name: "golden_table"
Priority:0
Key:"exact32" -> Value: Exact{.value = 10}
Key:"lpm32" -> Value: Lpm{.value = 0, .prefix_length = 0}
Key:"optional32" -> Value: Ternary{.value = 0, .mask = 0}
Key:"range32" -> Value: Range{.low = 0, .high = 4294967295}
Key:"ternary32" -> Value: Ternary{.value = 0, .mask = 0}

=== OUTPUT ===
All entries must satisfy:

In @entry_restriction of table 'golden_table'; at offset line 1, columns 23 to 40:
1 | exact32::value > 6 && exact32::value < 5;
  |                       ^^^^^^^^^^^^^^^^^^

But your entry does not.
>>>Entry Info<<<
Table Name: "golden_table"
Priority:0
Key:"exact32" -> Value: Exact{.value = 10}
Key:"lpm32" -> Value: Lpm{.value = 0, .prefix_length = 0}
Key:"optional32" -> Value: Ternary{.value = 0, .mask = 0}
Key:"range32" -> Value: Range{.low = 0, .high = 4294967295}
Key:"ternary32" -> Value: Ternary{.value = 0, .mask = 0}

### ReasonEntryViolatestConstraint Test ###################
=== INPUT ===
--- Constraint ---
exact32::value > 5;
exact32::value < 20;
exact32::value == 14;
--- Table Entry ---
Table Name: "golden_table"
Priority:0
Key:"exact32" -> Value: Exact{.value = 10}
Key:"lpm32" -> Value: Lpm{.value = 0, .prefix_length = 0}
Key:"optional32" -> Value: Ternary{.value = 0, .mask = 0}
Key:"range32" -> Value: Range{.low = 0, .high = 4294967295}
Key:"ternary32" -> Value: Ternary{.value = 0, .mask = 0}

=== OUTPUT ===
All entries must satisfy:

In @entry_restriction of table 'golden_table'; at offset line 3, columns 1 to 20:
  | exact32::value < 20;
3 | exact32::value == 14;
  | ^^^^^^^^^^^^^^^^^^^^

But your entry does not.
>>>Entry Info<<<
Table Name: "golden_table"
Priority:0
Key:"exact32" -> Value: Exact{.value = 10}
Key:"lpm32" -> Value: Lpm{.value = 0, .prefix_length = 0}
Key:"optional32" -> Value: Ternary{.value = 0, .mask = 0}
Key:"range32" -> Value: Range{.low = 0, .high = 4294967295}
Key:"ternary32" -> Value: Ternary{.value = 0, .mask = 0}

### ReasonEntryViolatestConstraint Test ###################
=== INPUT ===
--- Constraint ---
exact32::value > 0;
exact32::value > 7 || exact32::value == 5;
exact32::value == 9;
--- Table Entry ---
Table Name: "golden_table"
Priority:0
Key:"exact32" -> Value: Exact{.value = 10}
Key:"lpm32" -> Value: Lpm{.value = 0, .prefix_length = 0}
Key:"optional32" -> Value: Ternary{.value = 0, .mask = 0}
Key:"range32" -> Value: Range{.low = 0, .high = 4294967295}
Key:"ternary32" -> Value: Ternary{.value = 0, .mask = 0}

=== OUTPUT ===
All entries must satisfy:

golden_test.p4:3:1-19:
  | exact32::value > 7 || exact32::value == 5;
3 | exact32::value == 9;
  | ^^^^^^^^^^^^^^^^^^^

But your entry does not.
>>>Entry Info<<<
Table Name: "golden_table"
Priority:0
Key:"exact32" -> Value: Exact{.value = 10}
Key:"lpm32" -> Value: Lpm{.value = 0, .prefix_length = 0}
Key:"optional32" -> Value: Ternary{.value = 0, .mask = 0}
Key:"range32" -> Value: Range{.low = 0, .high = 4294967295}
Key:"ternary32" -> Value: Ternary{.value = 0, .mask = 0}

### ReasonEntryViolatestConstraint Test ###################
=== INPUT ===
--- Constraint ---
exact32::value > 0;
exact32::value < 42;
exact32::value < 20 -> exact32::value == 14;
--- Table Entry ---
Table Name: "golden_table"
Priority:0
Key:"exact32" -> Value: Exact{.value = 10}
Key:"lpm32" -> Value: Lpm{.value = 0, .prefix_length = 0}
Key:"optional32" -> Value: Ternary{.value = 0, .mask = 0}
Key:"range32" -> Value: Range{.low = 0, .high = 4294967295}
Key:"ternary32" -> Value: Ternary{.value = 0, .mask = 0}

=== OUTPUT ===
All entries must satisfy:

golden_test.p4:3:1-43:
  | exact32::value < 42;
3 | exact32::value < 20 -> exact32::value == 14;
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

But your entry does not.
>>>Entry Info<<<
Table Name: "golden_table"
Priority:0
Key:"exact32" -> Value: Exact{.value = 10}
Key:"lpm32" -> Value: Lpm{.value = 0, .prefix_length = 0}
Key:"optional32" -> Value: Ternary{.value = 0, .mask = 0}
Key:"range32" -> Value: Range{.low = 0, .high = 4294967295}
Key:"ternary32" -> Value: Ternary{.value = 0, .mask = 0}

### ReasonEntryViolatestConstraint Test ###################
=== INPUT ===
--- Constraint ---
exact32::value == 1 || exact32::value == 2;
!(exact32::value == 10 -> exact32::value == 10);
exact32::value == 3 || exact32::value == 4;
--- Table Entry ---
Table Name: "golden_table"
Priority:0
Key:"exact32" -> Value: Exact{.value = 10}
Key:"lpm32" -> Value: Lpm{.value = 0, .prefix_length = 0}
Key:"optional32" -> Value: Ternary{.value = 0, .mask = 0}
Key:"range32" -> Value: Range{.low = 0, .high = 4294967295}
Key:"ternary32" -> Value: Ternary{.value = 0, .mask = 0}

=== OUTPUT ===
All entries must not satisfy:

golden_test.p4:2:27-46:
  | exact32::value == 1 || exact32::value == 2;
2 | !(exact32::value == 10 -> exact32::value == 10);
  |                           ^^^^^^^^^^^^^^^^^^^^

But your entry does.
>>>Entry Info<<<
Table Name: "golden_table"
Priority:0
Key:"exact32" -> Value: Exact{.value = 10}
Key:"lpm32" -> Value: Lpm{.value = 0, .prefix_length = 0}
Key:"optional32" -> Value: Ternary{.value = 0, .mask = 0}
Key:"range32" -> Value: Range{.low = 0, .high = 4294967295}
Key:"ternary32" -> Value: Ternary{.value = 0, .mask = 0}

### ReasonEntryViolatestConstraint Test ###################
=== INPUT ===
--- Constraint ---
exact32::value == 80 || ternary32::value == 3096;
ternary32::mask == 255 && exact32::value == 3;
--- Table Entry ---
Table Name: "golden_table"
Priority:0
Key:"exact32" -> Value: Exact{.value = 10}
Key:"lpm32" -> Value: Lpm{.value = 0, .prefix_length = 0}
Key:"optional32" -> Value: Ternary{.value = 0, .mask = 0}
Key:"range32" -> Value: Range{.low = 0, .high = 4294967295}
Key:"ternary32" -> Value: Ternary{.value = 42, .mask = 64}

=== OUTPUT ===
All entries must satisfy:

In @entry_restriction of table 'golden_table'; at offset line 2, columns 1 to 22:
  | exact32::value == 80 || ternary32::value == 3096;
2 | ternary32::mask == 255 && exact32::value == 3;
  | ^^^^^^^^^^^^^^^^^^^^^^

But your entry does not.
>>>Entry Info<<<
Table Name: "golden_table"
Priority:0
Key:"exact32" -> Value: Exact{.value = 10}
Key:"lpm32" -> Value: Lpm{.value = 0, .prefix_length = 0}
Key:"optional32" -> Value: Ternary{.value = 0, .mask = 0}
Key:"range32" -> Value: Range{.low = 0, .high = 4294967295}
Key:"ternary32" -> Value: Ternary{.value = 42, .mask = 64}

### ReasonEntryViolatestConstraint Test ###################
=== INPUT ===
--- Constraint ---
(false || false) && (!(true -> true) && (false || false));
--- Table Entry ---
Table Name: "golden_table"
Priority:0
Key:"exact32" -> Value: Exact{.value = 10}
Key:"lpm32" -> Value: Lpm{.value = 0, .prefix_length = 0}
Key:"optional32" -> Value: Ternary{.value = 0, .mask = 0}
Key:"range32" -> Value: Range{.low = 0, .high = 4294967295}
Key:"ternary32" -> Value: Ternary{.value = 0, .mask = 0}

=== OUTPUT ===
All entries must not satisfy:

golden_test.p4:1:32-35:
1 | (false || false) && (!(true -> true) && (false || false));
  |                                ^^^^

But your entry does.
>>>Entry Info<<<
Table Name: "golden_table"
Priority:0
Key:"exact32" -> Value: Exact{.value = 10}
Key:"lpm32" -> Value: Lpm{.value = 0, .prefix_length = 0}
Key:"optional32" -> Value: Ternary{.value = 0, .mask = 0}
Key:"range32" -> Value: Range{.low = 0, .high = 4294967295}
Key:"ternary32" -> Value: Ternary{.value = 0, .mask = 0}

